hibike/hibike_simulator.py

```diff
@@ -1,133 +1,142 @@
+    
+    for read_thread in read_threads:
+        read_thread.start()
+    for write_thread in write_threads:
+        write_thread.start()
+
+    # the main thread reads instructions from statemanager and forwards them to the appropriate device write threads
+    while True:
+        instruction, args = pipeFromChild.recv()
+        if instruction == "enumerate_all":
+            for instruction_queue in instruction_queues:
+                instruction_queue.put(("ping", []))
+        elif instruction == "subscribe_device":
+            uid = args[0]
+            if uid in uid_to_index:
+                instruction_queues[uid_to_index[uid]].put(("subscribe", args))
+
+class DeviceWriteThread(threading.Thread):
+
+    def __init__(self, ser, instructionQueue, fake_device_queue):
+        self.ser = ser
+        self.queue = instructionQueue
+        self.fake_device_queue = fake_device_queue
+        super().__init__()
+
+    def run(self):
+        while True:
+            instruction = self.queue.get()
+            self.fake_device_queue.put(instruction)
+
+class DeviceReadThread(threading.Thread):
+
+    def __init__(self, ser, errorQueue, stateQueue, fake_device_queue):
+        self.ser = ser
+        self.errorQueue = errorQueue
+        self.stateQueue = stateQueue
+        self.fake_device_queue = fake_device_queue
+        self.delay = 0
+        self.params = []
+
+        self.fake_subscription_thread = FakeSubscriptionThread(fake_uids[self.ser], 0, [], self.fake_device_queue)
+        self.fake_subscription_thread.start()
+        super().__init__()
+
+    def run(self):
+        while True:
+            instruction, args = self.fake_device_queue.get()
+            res = None
+            if instruction == "ping":
+                uid, delay, params = fake_uids[self.ser], self.delay, self.params
+                uid_to_index[uid] = self.ser
+                res = ["device_subscribed", [uid, delay, params]]
+            elif instruction == "subscribe":
+                uid, delay, params = tuple(args)
+                uid_to_index[uid] = self.ser
+                res = ["device_subscribed", [uid, delay, params]]
+
+                self.fake_subscription_thread.quit = True
+                self.fake_subscription_thread.join()
+                self.fake_subscription_thread = FakeSubscriptionThread(uid, delay, params, self.fake_device_queue)
+                self.fake_subscription_thread.start()
+            elif instruction == "device_values":
+                res = [instruction, args]
+
+            self.stateQueue.put(res)
+
+class FakeSubscriptionThread(threading.Thread):
+
+    def __init__(self, uid, delay, params, fake_device_queue):
+        self.uid = uid
+        self.delay = delay
+        self.params = params
+        self.fake_device_queue = fake_device_queue
+        self.quit = False
+        super().__init__()
+
+    def run(self):
+        if self.delay != 0:
```

Comment:
This only supports subscribing. Unsubscribing in hibike is just subscribing with a delay of 0, which this also supports.